File: 2 Add Two Numbers/solution.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        $head = new ListNode(0);
        $current = $head;
        $carry = 0;

        while ($l1 != null || $l2 != null || $carry != 0) {
            $sum = $carry;

            if ($l1 != null) {
                $sum += $l1->val;
                $l1 = $l1->next;
            }

            if ($l2 != null) {
                $sum += $l2->val;
                $l2 = $l2->next;
            }

            $carry = (int) ($sum / 10);
            $current->next = new ListNode($sum % 10);
            $current = $current->next;
        }

        return $head->next;
    }
}

File: 2 Add Two Numbers/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut n1 = l1;
        let mut n2 = l2;
        let mut head = Some(Box::new(ListNode::new(0)));
        let mut current = head.as_mut();
        let mut carry = 0;

        while n1.is_some() || n2.is_some() || carry != 0 {
            let mut sum = carry;

            if let Some(node) = n1 {
                sum += node.val;
                n1 = node.next;
            }
            if let Some(node) = n2 {
                sum += node.val;
                n2 = node.next;
            }

            carry = sum / 10;
            
            if let Some(node) = current {
                node.next = Some(Box::new(ListNode::new(sum % 10)));
                current = node.next.as_mut();
            }
        }

        head.unwrap().next
    }
}